### PR TITLE
Compatibility with both Pydantic 1 and 2

### PR DESCRIPTION
**Pull request by hathawsh**
This PR makes the project compatible with both Pydantic 1 and 2. All Tox tests pass. Solved primarily by adding a `compat.py` shim that implements the changed symbols in 3 ways: Pydantic 1, and Pydantic 2, and type checking.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ include = ["openapi_pydantic/py.typed"]
 
 [tool.poetry.dependencies]
 python = "^3.8"
-pydantic = "^1.8.2"
+pydantic = "*"
```
Comment (mike-oakley) [inline pyproject.toml:26]: Lets keep the version a little more constrained - we don't want to support ancient versions (1.8 is 2 years old and removed a bunch of legacy bits).

```suggestion
pydantic = ">=1.8"
```

File: README.md [hunk 2]
```diff
@@ -37,7 +38,10 @@ open_api = OpenAPI(
         )
     },
 )
-print(open_api.json(by_alias=True, exclude_none=True, indent=2))
+if PYDANTIC_V2:
+    print(open_api.model_dump_json(by_alias=True, exclude_none=True, indent=2))
+else:
+    print(open_api.json(by_alias=True, exclude_none=True, indent=2))
```
Comment (mike-oakley) [inline README.md:44]: Not sure whether to complicate the README with the compat bits, or just assume v2 going forwards - what do you think?

File: openapi_pydantic/v3/v3_0_3/datatype.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 class DataType(str, enum.Enum):
     """Data type of an object."""
 
+    NULL = "null"
```
Comment (mike-oakley) [inline openapi_pydantic/v3/v3_0_3/datatype.py:7]: `null` isn't supported in v3.0.3 - https://swagger.io/specification/v3/

**Review by mike-oakley — COMMENTED** [review]
Awesome stuff! Few comments but looks good in general - going to be a bit of a pain to clean up when we remove 1.x support but can't think of a better way either.

Please could you add a section to the README explaining that both V1 and V2 are supported (and the general gist of the compat shim bit)?

**Comment by mike-oakley** [comment]
Closes #16

File: pyproject.toml (hunk 1 above)
Comment (hathawsh) [reply pyproject.toml]: Agreed and done.

File: README.md (hunk 2 above)
Comment (hathawsh) [reply README.md]: I agree, the conditions complicated the README. I changed the README so it assumes Pydantic 2, but I also added comments for Pydantic 1 users.

File: openapi_pydantic/v3/v3_0_3/datatype.py (hunk 3 above)
Comment (hathawsh) [reply openapi_pydantic/v3/v3_0_3/datatype.py]: Wow, learning that was a surprise! I added a test that exercised a nullable field. The test revealed that this package was breaking when generating OpenAPI 3.0.3 schemas containing nullable fields. Adding "null" as a data type was an obvious but wrong fix. I reverted that. Instead, I subclassed Pydantic's `GenerateJsonSchema` class to generate `nullable` fields (just for OpenAPI 3.0.3), and now all the tests pass again.

**Comment by hathawsh** [comment]
I added version compatibility notes to the README. Let me know if you spot anything else to improve this PR.

**Review by mike-oakley — APPROVED** [review]

**Comment by mike-oakley** [comment]
Thanks @hathawsh - just tagged 0.3.0 which should be on PyPI imminently 👍🏼

**Pull request merged**